### PR TITLE
Check workspace phase in ssh-gateway before connect to workspace

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

Check workspace phase in ssh-gateway before connect to workspace

## Related Issue(s)

Fixes IDE-109

## How to test

1. start a workspace using this branch
2. start a workspace using this preview env
3. upload your ssh public key, this step can make ssh connection url stable, otherwise you need get ownerToken everytime
4. in preview env workspace, do `dd if=/dev/random of=test.img bs=1M count=5000` it will created a large file in workspace folder, and run `gp stop` stop the workspace
5. during stopping, you can use ssh command trying to connect to workspace, it should failed
6. run `./dev/preview/portforward-monitoring-satellite.sh` in step 1 workspace, and open prometheus UI, search for `gitpod_ws_proxy_ssh_attempt_total` you should see there is a new error `WS_NOT_RUNNING`

image

7. repeat this test in preview env workspace Initializing, you should get same result.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - pd-ide-109</li>
	<li><b>🔗 URL</b> - <a href="https://pd-ide-109.preview.gitpod-dev.com/workspaces" target="_blank">pd-ide-109.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [x] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/ws-proxy/pkg/sshproxy/server.go [hunk 1]
```diff
@@ -249,6 +250,10 @@ func (s *Server) HandleConn(c net.Conn) {
 		ReportSSHAttemptMetrics(ErrWorkspaceNotFound)
 		return
 	}
+	if !wsInfo.IsRunning {
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply components/ws-proxy/pkg/sshproxy/server.go:253]: In such case the request in supervisor should be gracefully cancelled. You can check whether err is Cancelled then?

**Comment by akosyakov** [comment]
How to test it?

File: components/ws-proxy/pkg/sshproxy/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (akosyakov) [reply components/ws-proxy/pkg/sshproxy/server.go:253]: In this case state should be in stopping already?

**Comment by akosyakov** [comment]
ok, maybe move to draft for now then :)

**Review by Furisto — APPROVED** [review]

**Pull request merged**